src/python/dxpy/templating/templates/nextflow/src/nextflow.sh

```diff
@@ -75,29 +75,178 @@ on_exit() {
+  # clean up files of this session
+  else
+    echo "=== Execution complete — cache and working files will not be resumable"
+  fi
+
+  # remove .nextflow from the current folder /home/dnanexus/nextflow_execution
   rm -rf .nextflow
+  rm nxf_runtime.config
 
   # try uploading the log file if it is not empty
   if [[ -s $LOG_NAME ]]; then
-    mkdir ../nextflow_log
-    mv $LOG_NAME ../nextflow_log/$LOG_NAME || true
+    mkdir -p /home/dnanexus/out/nextflow_log
+    mv "$LOG_NAME" "/home/dnanexus/out/nextflow_log/$LOG_NAME" || true
   else
     echo "No nextflow log file available."
   fi
-  
+
   # upload the log file and published files if any
-  cd ..
-  if [[ -d ./nextflow_log || -n "$(ls -A ./output_files)" ]]; then
-    dx-upload-all-outputs --parallel || true
-  else
-    echo "No log file or output files has been generated."
-  fi
+  mkdir -p /home/dnanexus/out/published_files
+  find . -type f -newermt "$BEGIN_TIME" -exec mv {} /home/dnanexus/out/published_files/ \;
+
+  dx-upload-all-outputs --parallel --wait-on-close || echo "No log file or published files has been generated."
   # done
   exit $ret
 }
 
+get_resume_session_id() {
+  if [[ $resume == 'true' || $resume == 'last' ]]; then
+    # find the latest job run by applet with the same name
+    echo "Will try to find the session ID of the latest session run by $EXECUTABLE_NAME."
+    PREV_JOB_SESSION_ID=$(
+      dx api system findExecutions \
+        '{"state":["done","failed"],
+          "created": {"after":'$((($(date +%s) - 6 * 60 * 60 * 24 * 30) * 1000))'},
+          "project":"'$DX_PROJECT_CONTEXT_ID'",
+          "limit":1,
+          "includeSubjobs":false,
+          "describe":{"fields":{"properties":true}},
+          "properties":{
+            "nextflow_session_id":true,
+            "nextflow_preserve_cache":"true",
+            "nextflow_executable":"'$EXECUTABLE_NAME'"}}' 2>/dev/null |
+        jq -r '.results[].describe.properties.nextflow_session_id'
+    )
+
+    [[ -n $PREV_JOB_SESSION_ID ]] ||
+      dx-jobutil-report-error "Cannot find any jobs within the last 6 months to resume from. Please provide the exact sessionID for \"resume\" value or run without resume."
+  else
+    PREV_JOB_SESSION_ID=$resume
+  fi
+
+  valid_id_pattern='^\{?[A-Z0-9a-z]{8}-[A-Z0-9a-z]{4}-[A-Z0-9a-z]{4}-[A-Z0-9a-z]{4}-[A-Z0-9a-z]{12}\}?$'
+  [[ "$PREV_JOB_SESSION_ID" =~ $valid_id_pattern ]] ||
+    dx-jobutil-report-error "Invalid resume value. Please provide either \"true\", \"last\", or \"sessionID\". 
+    If a sessionID was provided, Nextflow cached content could not found under $DX_CACHEDIR/$PREV_JOB_SESSION_ID/. 
+    Please provide the exact sessionID for \"resume\" value or run without resume."
+
+  NXF_UUID=$PREV_JOB_SESSION_ID
+}
+
+restore_cache() {
+  # download latest cache.tar from $DX_CACHEDIR/$PREV_JOB_SESSION_ID/
+  PREV_JOB_CACHE_FILE=$(
+    dx api system findDataObjects \
+      '{"visibility": "either", 
+      "name":"cache.tar", 
```

Comment:
If you could also improve the formatting a little here?

Reply:
Done